### DISCUSSION TITLE
Nominatim search issue

### DISCUSSION THREAD

**Discussion by rgbsi-pp3** [osm-search/Nominatim#3584] [Q&A]
Hi all,

I am facing an error while searching place via nominatim. I have enabled the nominatim debug & the following is the error log:

postgres@forlinx:/run/media/mmcblk1p1/nominatim-project$ nominatim search --query Bangalore
2024-11-13 09:09:13: Using project directory: /run/media/mmcblk1p1/nominatim-project
2024-11-13 09:09:13,419 INFO sqlalchemy.engine.Engine select pg_catalog.version()
2024-11-13 09:09:13: select pg_catalog.version()
2024-11-13 09:09:13,420 INFO sqlalchemy.engine.Engine [raw sql] ()
2024-11-13 09:09:13: [raw sql] ()
2024-11-13 09:09:13,433 INFO sqlalchemy.engine.Engine select current_schema()
2024-11-13 09:09:13: select current_schema()
2024-11-13 09:09:13,434 INFO sqlalchemy.engine.Engine [raw sql] ()
2024-11-13 09:09:13: [raw sql] ()
2024-11-13 09:09:13,447 INFO sqlalchemy.engine.Engine show standard_conforming_strings
2024-11-13 09:09:13: show standard_conforming_strings
2024-11-13 09:09:13,448 INFO sqlalchemy.engine.Engine [raw sql] ()
2024-11-13 09:09:13: [raw sql] ()
2024-11-13 09:09:13,454 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-11-13 09:09:13: BEGIN (implicit)
2024-11-13 09:09:13,456 INFO sqlalchemy.engine.Engine SHOW server_version_num
2024-11-13 09:09:13: SHOW server_version_num
2024-11-13 09:09:13,457 INFO sqlalchemy.engine.Engine [generated in 0.00088s] ()
2024-11-13 09:09:13: [generated in 0.00088s] ()
2024-11-13 09:09:13,461 INFO sqlalchemy.engine.Engine SET jit_above_cost TO '-1'
2024-11-13 09:09:13: SET jit_above_cost TO '-1'
2024-11-13 09:09:13,462 INFO sqlalchemy.engine.Engine [generated in 0.00098s] ()
2024-11-13 09:09:13: [generated in 0.00098s] ()
2024-11-13 09:09:13,465 INFO sqlalchemy.engine.Engine SET max_parallel_workers_per_gather TO '0'
2024-11-13 09:09:13: SET max_parallel_workers_per_gather TO '0'
2024-11-13 09:09:13,466 INFO sqlalchemy.engine.Engine [generated in 0.00096s] ()
2024-11-13 09:09:13: [generated in 0.00096s] ()
2024-11-13 09:09:13,468 INFO sqlalchemy.engine.Engine COMMIT
2024-11-13 09:09:13: COMMIT
2024-11-13 09:09:13,502 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-11-13 09:09:13: BEGIN (implicit)
2024-11-13 09:09:13,513 INFO sqlalchemy.engine.Engine SELECT nominatim_properties.value
FROM nominatim_properties
WHERE nominatim_properties.property = %s
2024-11-13 09:09:13: SELECT nominatim_properties.value
FROM nominatim_properties
WHERE nominatim_properties.property = %s
2024-11-13 09:09:13,513 INFO sqlalchemy.engine.Engine [generated in 0.00129s] ('tokenizer',)
2024-11-13 09:09:13: [generated in 0.00129s] ('tokenizer',)
2024-11-13 09:09:13,608 INFO sqlalchemy.engine.Engine SELECT nominatim_properties.value
FROM nominatim_properties
WHERE nominatim_properties.property = %s
2024-11-13 09:09:13: SELECT nominatim_properties.value
FROM nominatim_properties
WHERE nominatim_properties.property = %s
2024-11-13 09:09:13,608 INFO sqlalchemy.engine.Engine [cached since 0.09646s ago] ('tokenizer_import_normalisation',)
2024-11-13 09:09:13: [cached since 0.09646s ago] ('tokenizer_import_normalisation',)
2024-11-13 09:09:13,779 INFO sqlalchemy.engine.Engine SELECT nominatim_properties.value
FROM nominatim_properties
WHERE nominatim_properties.property = %s
2024-11-13 09:09:13: SELECT nominatim_properties.value
FROM nominatim_properties
WHERE nominatim_properties.property = %s
2024-11-13 09:09:13,780 INFO sqlalchemy.engine.Engine [cached since 0.268s ago] ('tokenizer_import_transliteration',)
2024-11-13 09:09:13: [cached since 0.268s ago] ('tokenizer_import_transliteration',)
2024-11-13 09:09:13,858 INFO sqlalchemy.engine.Engine SELECT word.word_id, word.word_token, word.type, word.word, word.info
FROM word
WHERE word.word_token IN (%s)
2024-11-13 09:09:13: SELECT word.word_id, word.word_token, word.type, word.word, word.info
FROM word
WHERE word.word_token IN (%s)
2024-11-13 09:09:13,859 INFO sqlalchemy.engine.Engine [generated in 0.00162s] ('bangalore',)
2024-11-13 09:09:13: [generated in 0.00162s] ('bangalore',)
2024-11-13 09:09:13,957 INFO sqlalchemy.engine.Engine WITH searches AS
(SELECT anon_1.place_id AS place_id, anon_1.search_rank AS search_rank, anon_1.address_rank AS address_rank, anon_1.country_code AS country_code, anon_1.centroid AS centroid, anon_1.importance AS importance, anon_1.penalty AS penalty
FROM (SELECT anon_2.place_id AS place_id, anon_2.search_rank AS search_rank, anon_2.address_rank AS address_rank, anon_2.country_code AS country_code, anon_2.centroid AS centroid, anon_2.importance AS importance, anon_2.penalty AS penalty, first_value(anon_2.penalty - anon_2.importance) OVER (ORDER BY anon_2.penalty - anon_2.importance) AS min_penalty
FROM (SELECT search_name.place_id AS place_id, search_name.search_rank AS search_rank, search_name.address_rank AS address_rank, search_name.country_code AS country_code, search_name.centroid AS centroid, search_name.name_vector AS name_vector, search_name.nameaddress_vector AS nameaddress_vector, CASE WHEN (search_name.importance > %s) THEN search_name.importance ELSE %s - CAST(search_name.search_rank AS FLOAT) / %s END AS importance, %s + weigh_search(search_name.name_vector, %s, %s) AS penalty
FROM search_name
WHERE (search_name.name_vector @> %s) ORDER BY importance DESC
 LIMIT %s) AS anon_2) AS anon_1
WHERE anon_1.penalty - anon_1.importance < anon_1.min_penalty + %s)
 SELECT placex.place_id, placex.osm_type, placex.osm_id, placex.name, placex.class, placex.type, placex.address, placex.extratags, placex.housenumber, placex.postcode, placex.country_code, placex.wikipedia, placex.parent_place_id, placex.rank_address, placex.rank_search, placex.linked_place_id, placex.admin_level, ST_AsEWKB(placex.centroid) AS centroid, ST_AsEWKB(ST_Expand(placex.geometry, %s)) AS bbox, searches.importance, searches.penalty AS accuracy
FROM placex JOIN searches ON placex.place_id = searches.place_id
WHERE placex.linked_place_id IS NULL AND placex.indexed_status = %s ORDER BY searches.penalty - searches.importance, accuracy
 LIMIT %s
2024-11-13 09:09:13: WITH searches AS
(SELECT anon_1.place_id AS place_id, anon_1.search_rank AS search_rank, anon_1.address_rank AS address_rank, anon_1.country_code AS country_code, anon_1.centroid AS centroid, anon_1.importance AS importance, anon_1.penalty AS penalty
FROM (SELECT anon_2.place_id AS place_id, anon_2.search_rank AS search_rank, anon_2.address_rank AS address_rank, anon_2.country_code AS country_code, anon_2.centroid AS centroid, anon_2.importance AS importance, anon_2.penalty AS penalty, first_value(anon_2.penalty - anon_2.importance) OVER (ORDER BY anon_2.penalty - anon_2.importance) AS min_penalty
FROM (SELECT search_name.place_id AS place_id, search_name.search_rank AS search_rank, search_name.address_rank AS address_rank, search_name.country_code AS country_code, search_name.centroid AS centroid, search_name.name_vector AS name_vector, search_name.nameaddress_vector AS nameaddress_vector, CASE WHEN (search_name.importance > %s) THEN search_name.importance ELSE %s - CAST(search_name.search_rank AS FLOAT) / %s END AS importance, %s + weigh_search(search_name.name_vector, %s, %s) AS penalty
FROM search_name
WHERE (search_name.name_vector @> %s) ORDER BY importance DESC
 LIMIT %s) AS anon_2) AS anon_1
WHERE anon_1.penalty - anon_1.importance < anon_1.min_penalty + %s)
 SELECT placex.place_id, placex.osm_type, placex.osm_id, placex.name, placex.class, placex.type, placex.address, placex.extratags, placex.housenumber, placex.postcode, placex.country_code, placex.wikipedia, placex.parent_place_id, placex.rank_address, placex.rank_search, placex.linked_place_id, placex.admin_level, ST_AsEWKB(placex.centroid) AS centroid, ST_AsEWKB(ST_Expand(placex.geometry, %s)) AS bbox, searches.importance, searches.penalty AS accuracy
FROM placex JOIN searches ON placex.place_id = searches.place_id
WHERE placex.linked_place_id IS NULL AND placex.indexed_status = %s ORDER BY searches.penalty - searches.importance, accuracy
 LIMIT %s
2024-11-13 09:09:13,958 INFO sqlalchemy.engine.Engine [generated in 0.00280s] (0, 0.40001, 75, 0.0, '[[0.0,[241022]]]', 0.5, [5452], 10000, 0.5, 0, 0, 20)
2024-11-13 09:09:13: [generated in 0.00280s] (0, 0.40001, 75, 0.0, '[[0.0,[241022]]]', 0.5, [5452], 10000, 0.5, 0, 0, 20)
2024-11-13 09:09:15,285 INFO sqlalchemy.engine.Engine SELECT CAST(anon_1.value ->> %s AS INTEGER) AS src_place_id, placex.place_id, placex.osm_type, placex.osm_id, placex.name, placex.class, placex.type, placex.extratags, placex.admin_level, place_addressline.fromarea, CASE WHEN (placex.rank_address = %s) THEN %s ELSE placex.rank_address END AS rank_address, place_addressline.distance, placex.country_code, placex.postcode
FROM json_array_elements(%s) AS anon_1 JOIN place_addressline ON place_addressline.place_id = CAST((anon_1.value ->> %s) AS INTEGER) OR place_addressline.place_id = CAST((anon_1.value ->> %s) AS INTEGER) JOIN placex ON place_addressline.address_place_id = placex.place_id ORDER BY src_place_id, rank_address DESC, place_addressline.place_id = CAST((anon_1.value ->> %s) AS INTEGER) DESC, CASE WHEN coalesce(avals(placex.name) && ARRAY(SELECT * FROM json_array_elements_text(anon_1.value -> %s)), false) THEN %s WHEN place_addressline.isaddress THEN %s WHEN (place_addressline.fromarea AND ST_Contains(placex.geometry, ST_GeomFromEWKT(CAST(anon_1.value ->> %s AS VARCHAR)))) THEN %s ELSE %s END DESC, place_addressline.fromarea DESC, place_addressline.distance DESC, placex.rank_search DESC
2024-11-13 09:09:15: SELECT CAST(anon_1.value ->> %s AS INTEGER) AS src_place_id, placex.place_id, placex.osm_type, placex.osm_id, placex.name, placex.class, placex.type, placex.extratags, placex.admin_level, place_addressline.fromarea, CASE WHEN (placex.rank_address = %s) THEN %s ELSE placex.rank_address END AS rank_address, place_addressline.distance, placex.country_code, placex.postcode
FROM json_array_elements(%s) AS anon_1 JOIN place_addressline ON place_addressline.place_id = CAST((anon_1.value ->> %s) AS INTEGER) OR place_addressline.place_id = CAST((anon_1.value ->> %s) AS INTEGER) JOIN placex ON place_addressline.address_place_id = placex.place_id ORDER BY src_place_id, rank_address DESC, place_addressline.place_id = CAST((anon_1.value ->> %s) AS INTEGER) DESC, CASE WHEN coalesce(avals(placex.name) && ARRAY(SELECT * FROM json_array_elements_text(anon_1.value -> %s)), false) THEN %s WHEN place_addressline.isaddress THEN %s WHEN (place_addressline.fromarea AND ST_Contains(placex.geometry, ST_GeomFromEWKT(CAST(anon_1.value ->> %s AS VARCHAR)))) THEN %s ELSE %s END DESC, place_addressline.fromarea DESC, place_addressline.distance DESC, placex.rank_search DESC
2024-11-13 09:09:15,286 INFO sqlalchemy.engine.Engine [generated in 0.00290s] ('pid', 11, 5, '[{"pid": 7819670, "lid": 7819670, "names": ["560001", "IN-KA"], "c": "SRID=4326;POINT(77.590082 12.9767936)"}, {"pid": 7816449, "lid": 7816449, "name ... (861 characters truncated) ... "c": "SRID=4326;POINT(77.70540295967255 12.994626449999998)"}, {"pid": 8012027, "lid": 8012027, "names": [], "c": "SRID=4326;POINT(77.583333 13.0)"}]', 'pid', 'lid', 'pid', 'names', 2, 0, 'c', 1, -1)
2024-11-13 09:09:15: [generated in 0.00290s] ('pid', 11, 5, '[{"pid": 7819670, "lid": 7819670, "names": ["560001", "IN-KA"], "c": "SRID=4326;POINT(77.590082 12.9767936)"}, {"pid": 7816449, "lid": 7816449, "name ... (861 characters truncated) ... "c": "SRID=4326;POINT(77.70540295967255 12.994626449999998)"}, {"pid": 8012027, "lid": 8012027, "names": [], "c": "SRID=4326;POINT(77.583333 13.0)"}]', 'pid', 'lid', 'pid', 'names', 2, 0, 'c', 1, -1)
Traceback (most recent call last):
  File "/usr/bin/nominatim", line 13, in <module>
    osm2pgsql_path='/usr/lib/nominatim/osm2pgsql'))
  File "/usr/lib/nominatim/lib-python/nominatim/cli.py", line 228, in nominatim
    return get_set_parser().run(**kwargs)
  File "/usr/lib/nominatim/lib-python/nominatim/cli.py", line 121, in run
    return args.command.run(args)
  File "/usr/lib/nominatim/lib-python/nominatim/clicmd/api.py", line 118, in run
    results = api.search(args.query, **params)
  File "/usr/lib/nominatim/lib-python/nominatim/api/core.py", line 748, in search
    self._async_api.search(query, **params))
  File "usr/lib/python3.7/asyncio/base_events.py", line 584, in run_until_complete
  File "/usr/lib/nominatim/lib-python/nominatim/api/core.py", line 261, in search
    return await geocoder.lookup(phrases)
  File "/usr/lib/nominatim/lib-python/nominatim/api/search/geocoder.py", line 216, in lookup
    await add_result_details(self.conn, results, self.params)
  File "/usr/lib/nominatim/lib-python/nominatim/api/results.py", line 451, in add_result_details
    await complete_address_details(conn, results)
  File "/usr/lib/nominatim/lib-python/nominatim/api/results.py", line 627, in complete_address_details
    for row in await conn.execute(sql):
  File "/usr/lib/nominatim/lib-python/nominatim/api/connection.py", line 63, in execute
    return await asyncio.wait_for(self.connection.execute(sql, params), self.query_timeout)
  File "usr/lib/python3.7/asyncio/tasks.py", line 423, in wait_for
concurrent.futures._base.TimeoutError

I have given the following configurations in postgresql.conf (Recommended by nominatim docs):
shared_buffers = 2GB
maintenance_work_mem = (10GB)
autovacuum_work_mem = 2GB
work_mem = (50MB)
synchronous_commit = off
max_wal_size = 1GB
checkpoint_timeout = 60min
checkpoint_completion_target = 0.9
random_page_cost = 1.0
wal_level = minimal
max_wal_senders = 0
autovacuum_max_workers = 1

kindly provide me the solution for this error as I got stuck with this issue. Thanks in advance.

**Comment by mtmail** [comment]
This seems to be the relevant line
> await asyncio.wait_for(self.connection.execute(sql, params), self.query_timeout)

Do you have an `.env` file in your project directory and is `NOMINATIM_DATABASE_DSN` set in there? Or somewhere else? By default Nominatim tries to connect to a database called `nominatim` on localhost. https://nominatim.org/release-docs/latest/customize/Settings

Can you login to the database with `psql` or do you see errors there, too?

Did the Nominatim data import finish completely? What does `nominatim admin --check-database` print?

**Comment by rgbsi-pp3** [comment]
.env file :
NOMINATIM_DATABASE_DSN="pgsql:dbname=nominatim;user=postgres"
NOMINATIM_QUERY_TIMEOUT=10
NOMINATIM_REQUEST_TIMEOUT=10
NOMINATIM_LOG_DB=no
NOMINATIM_DEBUG_SQL=yes

-----------------------------------------------------------------------------------------------------------------

postgres@forlinx:/run/media/mmcblk1p1/nominatim-project$ nominatim admin --check-database
2024-11-14 06:00:59: Using project directory: /run/media/mmcblk1p1/nominatim-project
2024-11-14 06:00:59: Checking database
Checking database connection ... OK
Checking database_version matches Nominatim software version ... OK
Checking for placex table ... OK
Checking for placex content ... OK
Checking that tokenizer works ... OK
Checking for wikipedia/wikidata data ... not applicable
Checking indexing status ... WARNING

 The indexing didn't finish. 888 entries are not yet indexed.

 To index the remaining entries, run:               Database is marked frozen, it cannot be updated.
Low counts of unindexed places are fine.

Checking that database indexes are complete ... OK
Checking that all database indexes are valid ... OK
Checking TIGER external data table. ... not applicable

----------------------------------------------------------------------------------------------------------------
postgres@forlinx:~$ psql -d nominatim
psql (14.12)
Type "help" for help.

nominatim=# \l
                                  List of databases
   Name    |  Owner   | Encoding |   Collate   |    Ctype    |   Access privileges
-----------+----------+----------+-------------+-------------+-----------------------
 nominatim | postgres | UTF8     | en_US.UTF-8 | en_US.UTF-8 | =Tc/postgres         +
           |          |          |             |             | postgres=CTc/postgres
 postgres  | postgres | UTF8     | en_US.UTF-8 | en_US.UTF-8 |
 template0 | postgres | UTF8     | en_US.UTF-8 | en_US.UTF-8 | =c/postgres          +
           |          |          |             |             | postgres=CTc/postgres
 template1 | postgres | UTF8     | en_US.UTF-8 | en_US.UTF-8 | =c/postgres          +
           |          |          |             |             | postgres=CTc/postgres
(4 rows)

I am able to login to nominatim database by psql commands @mtmail

**Comment by rgbsi-pp3** [comment]
The indexing didn't finish. 888 entries are not yet indexed.

I hope this is not the issue, as I am running nominatim with my other ubuntu machine with same indexing and I am able to get search results.

I am replicating the database to my target device using pg_dump & pg_restore with same postgresql & nominatim version as my ubuntu system.

The logs that I have shared above are from my target device in which timeout error occurs on nominatim search

**Reply by mtmail** [reply]
The 888 unindexed places are fine, that can happen with any import.

"Database is marked frozen" (the code checks if a database table named 'place' exists). Is that expected? I mean did you run 'nominatim freeze' after the data import (see https://nominatim.org/release-docs/latest/admin/Import/#dropping-data-required-for-dynamic-updates) ?

Did you use a flatnode file during import? If you did then you also need to copy it to the new server. (see https://nominatim.org/release-docs/latest/admin/Import/#flatnode-files)

I'm not really sure why the timeout happens. It might be missing indexes in the target Postgresql installation. pg_restore usually recreates them, but I don't know if pg_dump dumps them by default.

**Reply by rgbsi-pp3** [reply]
I did nominatim freeze. I am working on memory constrain target device. In nominatim docs I read that by freezing the database will reduce the database size, as most of the data are for just updates which I think is not needed for my application.

I didn't use any flatnode file. I am not aware of that file usage

**Reply by mtmail** [reply]
Using the freezing/readonly feature if fine. I was just wondering if it was on purpose, because if not then it would've given us a clue that some data might be missing.
How memory constrained is the server (CPU cores, RAM, type of harddrive)? Which country or regions did you originally import?

**Reply by rgbsi-pp3** [reply]
Basically, it is a stand-alone device that has offline map Application
I am using a Quad-Core Cortex-A53 CPU with 2GB DDR3L RAM
I am using SD Card with 512 GB storage for having Database & Swapfile utilization

I am importing latest India region(.pbf) that I got from Geofabrik website.

**Reply by mtmail** [reply]
The postgresql.conf configuration settings you posted are for a 128GB RAM machine (https://nominatim.org/release-docs/latest/admin/Installation/#tuning-the-postgresql-database). Try reducing the buffers. 2GB and 10 year old CPU architecture is though for such larger country, I'm not sure you can ever get good performance

**Reply by rgbsi-pp3** [reply]
I have created a new database without any indexing issue & without freezing the database. But still the timeout issue exists.

root@forlinx:/run/media/mmcblk1p1/nominatim-project# nominatim admin --check-database
2024-11-21 05:03:38: Using project directory: /run/media/mmcblk1p1/nominatim-project
2024-11-21 05:03:38: Checking database
Checking database connection ... OK
Checking database_version matches Nominatim software version ... OK
Checking for placex table ... OK
Checking for placex content ... OK
Checking that tokenizer works ... OK
Checking for wikipedia/wikidata data ... WARNING

Wikipedia/Wikidata importance tables missing.
Quality of search results may be degraded. Reverse geocoding is unaffected.
See https://nominatim.org/release-docs/latest/admin/Import/#wikipediawikidata-rankings

Checking indexing status ... OK
Checking that database indexes are complete ... OK
Checking that all database indexes are valid ... OK

-----------------------------------------------------------------------------------------------------------------------------------------

I am new to postgresql, so kindly suggest me postgresql.conf file settings for my hardware configurations.
Right now, I am using a swapfile to match those configurations above mentioned postgresql.conf configurations.

Is the timeout is due to low device RAM or any other hardware constrain?

**Reply by mtmail** [reply]
Most likely hardware constraints. You’ll meed to do own experimentation with postgresql setting specific to your setup.

**Comment by rgbsi-pp3** [comment]
I have manually searched for place "bangalore" via nominatim psql database.. The Bellow log is the response that I got within seconds: 

postgres@forlinx:/run/media/mmcblk1p1/nominatim-project$ psql "dbname=nominatim host=/tmp"
nominatim=# SELECT place_id, osm_type, osm_id, class, type, admin_level, name, ST_AsText(geometry)
nominatim-# FROM placex
nominatim-# WHERE name->'name' ILIKE '%Bangalore%'
nominatim-# LIMIT 10;
 place_id | osm_type |   osm_id    |  class   |      type       | admin_level |                                                                       name                                                                        |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 st_astext                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     
----------+----------+-------------+----------+-----------------+-------------+---------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  8262397 | W        |   331814099 | amenity  | public_building |          15 | "name"=>"Bangalore One", "name:kn"=>"ಬೆಂಗಳೂರು ಓನ್"                                                                          | POLYGON((77.6372547 12.9813192,77.6372992 12.9813092,77.6373085 12.9813487,77.637264 12.9813586,77.6372547 12.9813192))
  8264235 | W        |   143384387 | historic | castle          |          15 | "name"=>"Bangalore Palace", "name:kn"=>"ಬೆಂಗಳೂರು ಅರಮನೆ", "name:ml"=>"ബെംഗളൂരു കൊട്ടാരം" | POLYGON((77.5915948 12.9983437,77.5916499 12.9982914,77.5916133 12.9982513,77.5916854 12.998188,77.5917259 12.998233,77.5917581 12.9982042,77.591885 12.9983303,77.591903 12.9983218,77.5919106 12.9983226,77.5919175 12.9983258,77.5919226 12.9983309,77.5919258 12.9983377,77.5919265 12.9983451,77.5919235 12.9983542,77.5919162 12.9983686,77.591957 12.9984148,77.5919627 12.9984118,77.591969 12.998411,77.5919725 12.9984114,77.5919784 12.9984139,77.5919831 12.9984181,77.5919861 12.9984237,77.5919867 12.9984313,77.5919848 12.9984373,77.5919809 12.9984423,77.5919906 12.9984541,77.5920015 12.9984461,77.592029 12.9984301,77.5920541 12.9984246,77.592086 12.9984275,77.5921156 12.9984412,77.592138 12.9984635,77.592152 12.9984953,77.592152 12.9985329,77.5921414 12.9985595,77.5921207 12.9985833,77.592147 12.9986093,77.592195 12.9985624,77.592242 12.9986075,77.5922098 12.9986362,77.5922115 12.9986376,77.5922179 12.9986325,77.5922248 12.99863,77.592233 12.9986297,77.5922409 12.9986322,77.5922474 12.9986372,77.5922518 12.9986442,77.5922535 12.9986521,77.5922522 12.9986601,77.5922483 12.9986672,77.5923804 12.9988236,77.5922114 12.9989621,77.5921741 12.9989904,77.5921777 12.9989975,77.5921786 12.9990054,77.5921751 12.9990161,77.5921696 12.999022,77.5921624 12.9990258,77.5921543 12.9990269,77.5921463 12.9990252,77.5921404 12.9990219,77.592135 12.9990159,77.592132 12.9990085,77.5921318 12.9990005,77.5921344 12.9989929,77.5921384 12.9989877,77.5919992 12.9988365,77.5919527 12.998873,77.591905 12.9988247,77.5918489 12.9988743,77.5917859 12.998808,77.5918434 12.9987585,77.5917983 12.9987112,77.5918118 12.9986877,77.5918118 12.9986642,77.5918189 12.9986511,77.5918305 12.9986407,77.5918574 12.9986355,77.5918976 12.9986093,77.5918622 12.9985724,77.5919056 12.9985335,77.5918064 12.9984342,77.591743 12.9984943,77.5915948 12.9983437))
  8264759 | N        | 11472852219 | amenity  | driving_school  |          15 | "name"=>"Bangalore Driving School"                                                                                                                | POINT(77.619535 12.9997999)
  8265024 | N        |  1279246451 | office   | government      |          15 | "name"=>"Bangalore One", "name:kn"=>"ಬೆಂಗಳೂರು ಓನ್"                                                                          | POINT(77.6377482 12.9136474)
  8265462 | N        | 11969574174 | office   | government      |          15 | "name"=>"Bangalore One Citizen Service Centre"                                                                                                    | POINT(77.567461 12.9970088)
  8267131 | N        |  3751686636 | office   | society         |          15 | "name"=>"Bangalore Association for Science Education"                                                                                             | POINT(77.5895991 12.9848736)
  8267278 | W        |    37869985 | leisure  | sports_centre   |          15 | "name"=>"Bangalore Race Course"                                                                                                                   | POLYGON((77.5767188 12.9822315,77.576733 12.9819496,77.5767882 12.9816919,77.5768952 12.9814582,77.5770788 12.9812372,77.5772754 12.9810602,77.5774525 12.9809484,77.5777378 12.9808682,77.5779868 12.9808223,77.5784918 12.9807801,77.578961 12.9807547,77.5793705 12.980755,77.5804979 12.9807558,77.5825229 12.9807638,77.5831358 12.9807662,77.584899 12.9806046,77.5850995 12.9805802,77.5851696 12.9808833,77.5845145 12.9810026,77.5845846 12.9812981,77.5847105 12.9815025,77.5848194 12.9819951,77.5848239 12.9824202,77.5848181 12.9826312,77.5844109 12.983222,77.584432 12.9837064,77.584319 12.9837452,77.583339 12.9841084,77.5830334 12.9841652,77.5823613 12.9841786,77.5822115 12.9841465,77.5819854 12.9840627,77.581875 12.9839879,77.5818222 12.9839232,77.5811592 12.983599,77.5795788 12.9837209,77.5789658 12.9837795,77.5786875 12.9837944,77.5784705 12.9837687,77.5780931 12.9836747,77.5777491 12.9835639,77.5774293 12.9834404,77.5771443 12.9832291,77.5769398 12.9829176,77.5767937 12.9826071,77.5767188 12.9822315))
  8267630 | W        |   373566738 | building | train_station   |          15 | "name"=>"Bangalore Junction", "name:kn"=>"ಕ್ರಾಂತಿವೀರ ಸಂಗೊಳ್ಳಿ ರಾಯಣ್ಣ"                             | POLYGON((77.5693639 12.9775061,77.5694256 12.9774963,77.5694706 12.9774891,77.5695105 12.977709,77.5695363 12.9778623,77.5695878 12.9779265,77.5696164 12.9779216,77.5696357 12.9780294,77.5696755 12.9782652,77.5696854 12.9783176,77.5696877 12.9783485,77.569656 12.9783933,77.5696585 12.9784282,77.5696225 12.9784332,77.5696776 12.9787765,77.569702 12.9789245,77.5695968 12.978941,77.5693639 12.9775061))
  8267745 | W        |  1069218265 | amenity  | school          |          15 | "name"=>"Bangalore Education Society"                                                                                                             | POLYGON((77.5684127 12.9991976,77.5684193 12.9987684,77.5689087 12.9987755,77.5689021 12.9992047,77.5684127 12.9991976))
  8268000 | N        |   806588788 | railway  | stop            |          15 | "ref"=>"SBC:10", "name"=>"Bangalore City Junction", "name:kn"=>"ಬೆಂಗಳೂರು ನಗರ"                                               | POINT(77.5678774 12.9784008)
(10 rows)

------------------------------------------------------------------------------------------------------------------------------------------------------
I have modified postgresql.conf file according to my device RAM as follows:
shared_buffers = 512MB
maintenance_work_mem = 128mb
autovacuum_work_mem = 64MB
work_mem = 4MB

But still the timeout issue persists..

---------------------------------------------------------------------------------------------------------------------------------------------------

postgres@forlinx:/run/media/mmcblk1p1/nominatim-project$ cat ../Database/logfile
2024-11-25 10:29:41.323 UTC [2840] LOG:  starting PostgreSQL 14.12 on aarch64-buildroot-linux-gnu, compiled by aarch64-linux-gnu-gcc (Linaro GCC 7.4-2019.02) 7.4.1 20181213 [linaro-7.4-2019.02 revision 56ec6f6b99cc167ff0c2f8e1a2eed33b1edc85d4], 64-bit
2024-11-25 10:29:41.324 UTC [2840] LOG:  listening on IPv4 address "127.0.0.1", port 5432
2024-11-25 10:29:41.335 UTC [2840] LOG:  listening on Unix socket "/tmp/.s.PGSQL.5432"
2024-11-25 10:29:41.374 UTC [2841] LOG:  database system was shut down at 2024-11-25 10:29:24 UTC
2024-11-25 10:29:41.391 UTC [2840] LOG:  database system is ready to accept connections
2024-11-25 10:34:52.081 UTC [2853] LOG:  could not send data to client: Broken pipe
2024-11-25 10:34:52.081 UTC [2853] STATEMENT:  SELECT CAST(anon_1.value ->> $1::varchar AS INTEGER) AS src_place_id, placex.place_id, placex.osm_type, placex.osm_id, placex.name, placex.class, placex.type, placex.extratags, placex.admin_level, place_addressline.fromarea, CASE WHEN (placex.rank_address = $2::integer) THEN $3::integer ELSE placex.rank_address END AS rank_address, place_addressline.distance, placex.country_code, placex.postcode
        FROM json_array_elements($4::json) AS anon_1 JOIN place_addressline ON place_addressline.place_id = CAST((anon_1.value ->> $5::varchar) AS INTEGER) OR place_addressline.place_id = CAST((anon_1.value ->> $6::varchar) AS INTEGER) JOIN placex ON place_addressline.address_place_id = placex.place_id ORDER BY src_place_id, rank_address DESC, place_addressline.place_id = CAST((anon_1.value ->> $7::varchar) AS INTEGER) DESC, CASE WHEN coalesce(avals(placex.name) && ARRAY(SELECT * FROM json_array_elements_text(anon_1.value -> $8::varchar)), false) THEN $9::integer WHEN place_addressline.isaddress THEN $10::integer WHEN (place_addressline.fromarea AND ST_Contains(placex.geometry, ST_GeomFromEWKT(CAST(anon_1.value ->> $11::varchar AS VARCHAR)))) THEN $12::integer ELSE $13::integer END DESC, place_addressline.fromarea DESC, place_addressline.distance DESC, placex.rank_search DESC
2024-11-25 10:34:52.082 UTC [2853] FATAL:  connection to client lost
2024-11-25 10:34:52.082 UTC [2853] STATEMENT:  SELECT CAST(anon_1.value ->> $1::varchar AS INTEGER) AS src_place_id, placex.place_id, placex.osm_type, placex.osm_id, placex.name, placex.class, placex.type, placex.extratags, placex.admin_level, place_addressline.fromarea, CASE WHEN (placex.rank_address = $2::integer) THEN $3::integer ELSE placex.rank_address END AS rank_address, place_addressline.distance, placex.country_code, placex.postcode
        FROM json_array_elements($4::json) AS anon_1 JOIN place_addressline ON place_addressline.place_id = CAST((anon_1.value ->> $5::varchar) AS INTEGER) OR place_addressline.place_id = CAST((anon_1.value ->> $6::varchar) AS INTEGER) JOIN placex ON place_addressline.address_place_id = placex.place_id ORDER BY src_place_id, rank_address DESC, place_addressline.place_id = CAST((anon_1.value ->> $7::varchar) AS INTEGER) DESC, CASE WHEN coalesce(avals(placex.name) && ARRAY(SELECT * FROM json_array_elements_text(anon_1.value -> $8::varchar)), false) THEN $9::integer WHEN place_addressline.isaddress THEN $10::integer WHEN (place_addressline.fromarea AND ST_Contains(placex.geometry, ST_GeomFromEWKT(CAST(anon_1.value ->> $11::varchar AS VARCHAR)))) THEN $12::integer ELSE $13::integer END DESC, place_addressline.fromarea DESC, place_addressline.distance DESC, placex.rank_search DESC


The above is my the postgresql databse log.. As, I use nominatim search there is client broken pipe error..
2024-11-25 10:34:52.081 UTC [2853] LOG:  could not send data to client: Broken pipe

**Comment by rgbsi-pp3** [comment]
@mtmail  Since I am able to get response direclty from postgresql server, could you provide me the SQL query for search to be executed that provides the result of nominatim search?

**Reply by mtmail** [reply]
You can run `nominatim search --query Bangalore --format debug` to see the SQL statements. (Requires Nominatim version 4.5)
